### ISSUE TITLE
LT02/09 "fixes not applied, as it would result in an unparsable file" error

### ISSUE THREAD

**Issue by ahwitz** [sqlfluff/sqlfluff#7529]
### Search before asking

- [x] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

(hard to determine if it's a full dupe with existing LT02/09 tickets)

Got the "please report this as a bug" error for those two rules. If working on this is straightforward, I've probably got availability to learn.

### Expected Behaviour

Should build.

### Observed Behaviour

Doesn't build.

### How to reproduce

```sql
WITH GROUPED_IDS AS (
    SELECT ID
        , FIRST_VALUE(ID) OVER (
            PARTITION BY GROUP_KEY ORDER BY (DATA_SOURCE = '[removed for bug submission]', DATA_SOURCE) DESC
        ) AS GROUPED_ID
    FROM BASE_GROUPED
)

, DIAGNOSIS_CODES AS (
    SELECT DISTINCT GI.GROUPED_ID AS ID
        , DC.VALUE::VARCHAR AS DIAG_CODE
        , IFF(CONTAINS(DIAG_CODE, ','), NULL, DIAG_DESCRIPTION) AS DIAG_DESCRIPTION
    FROM GROUPED_IDS AS GI
    INNER JOIN ADT_RAW AS AR
        ON GI.ID = AR.ID
            , LATERAL FLATTEN(INPUT => SPLIT(AR.DIAG_CODE, ','), OUTER => TRUE) AS DC
)

SELECT *
FROM ADT_RAW
```

SQL does not compile, but this was the smallest subset of a giant file that I could base case to get the error. Requires both CTEs in place.

### Dialect

Snowflake/DBT, but this doesn't have any active Jinja.

### Version

Unknown, but whatever version is baked into DBT Cloud's "Latest" version - sorry this is useless, but they make it amazingly hard to find any more info than this.

### Configuration

```
[sqlfluff]
dialect = snowflake
templater = dbt
max_line_length = 120
# Number of passes to run before admitting defeat
runaway_limit = 5

# CPU processes to use while linting.
processes = 2

# can either be autodetect or a valid encoding e.g. utf-8, utf-8-sig
encoding = autodetect

# Ignore inline overrides (e.g. to test if still required)
disable_noqa = False

# Allow fix to run on files, even if they contain parsing errors
# Note altering this is NOT RECOMMENDED as can corrupt SQL
fix_even_unparsable = False

# Very large files can make the parser effectively hang.
# The more efficient check is the _byte_ limit check which
# is enabled by default. The previous _character_ limit check
# is still present for backward compatibility. This will be
# removed in a future version.
# Set either to 0 to disable.
large_file_skip_char_limit = 0
large_file_skip_byte_limit = 40000

# NOTE: Templater Variant rendering should currently be considered EXPERIMENTAL.
render_variant_limit = 1

# If using the dbt templater, we recommend setting the project dir.
[sqlfluff:templater:dbt]
project_dir = ./

[sqlfluff:indentation]
# See https://docs.sqlfluff.com/en/stable/perma/indent_locations.html
indent_unit = space
tab_space_size = 4
indented_joins = False
indented_ctes = False
indented_using_on = True
indented_on_contents = False
indented_then = True
indented_then_contents = True
allow_implicit_indents = True
template_blocks_indent = True

# If comments are found at the end of long lines, we default to moving
# them to the line _before_ their current location as the convention is
# that a comment precedes the line it describes. However if you prefer
# comments moved _after_, this configuration setting can be set to "after".
trailing_comments = before

# To exclude comment lines from indentation entirely set this to "True".
ignore_comment_lines = False

# Layout configuration
# See https://docs.sqlfluff.com/en/stable/perma/layout_spacing.html
[sqlfluff:layout:type:comma]
spacing_before = touch
line_position = leading

[sqlfluff:layout:type:binary_operator]
spacing_within = touch
line_position = leading

[sqlfluff:layout:type:column_path_operator]
spacing_within = touch
line_position = leading

[sqlfluff:layout:type:statement_terminator]
spacing_before = touch
line_position = leading

[sqlfluff:layout:type:end_of_file]
spacing_before = touch

[sqlfluff:layout:type:set_operator]
line_position = alone:strict

[sqlfluff:layout:type:start_bracket]
spacing_after = touch

[sqlfluff:layout:type:end_bracket]
spacing_before = touch

[sqlfluff:layout:type:start_square_bracket]
spacing_after = touch

[sqlfluff:layout:type:end_square_bracket]
spacing_before = touch

[sqlfluff:layout:type:start_angle_bracket]
spacing_after = touch

[sqlfluff:layout:type:end_angle_bracket]
spacing_before = touch

[sqlfluff:layout:type:casting_operator]
spacing_before = touch
spacing_after = touch:inline

[sqlfluff:layout:type:slice]
spacing_before = touch
spacing_after = touch

[sqlfluff:layout:type:dot]
spacing_before = touch
spacing_after = touch

[sqlfluff:layout:type:comparison_operator]
spacing_within = touch
line_position = leading

[sqlfluff:layout:type:assignment_operator]
spacing_within = touch
line_position = leading

[sqlfluff:layout:type:object_reference]
spacing_within = touch:inline

[sqlfluff:layout:type:numeric_literal]
spacing_within = touch:inline

[sqlfluff:layout:type:sign_indicator]
spacing_after = touch:inline

[sqlfluff:layout:type:tilde]
spacing_after = touch:inline

[sqlfluff:layout:type:function_name]
spacing_within = touch:inline

[sqlfluff:layout:type:function_contents]
spacing_before = touch:inline

[sqlfluff:layout:type:function_parameter_list]
spacing_before = touch:inline

[sqlfluff:layout:type:array_type]
spacing_within = touch:inline

[sqlfluff:layout:type:typed_array_literal]
spacing_within = touch

[sqlfluff:layout:type:sized_array_type]
spacing_within = touch

[sqlfluff:layout:type:struct_type]
spacing_within = touch:inline

[sqlfluff:layout:type:bracketed_arguments]
spacing_before = touch:inline

[sqlfluff:layout:type:match_condition]
spacing_within = touch:inline

[sqlfluff:layout:type:typed_struct_literal]
spacing_within = touch

[sqlfluff:layout:type:semi_structured_expression]
spacing_within = touch:inline
spacing_before = touch:inline

[sqlfluff:layout:type:array_accessor]
spacing_before = touch:inline

[sqlfluff:layout:type:colon]
spacing_before = touch

[sqlfluff:layout:type:colon_delimiter]
spacing_before = touch
spacing_after = touch

[sqlfluff:layout:type:path_segment]
spacing_within = touch

[sqlfluff:layout:type:sql_conf_option]
spacing_within = touch

[sqlfluff:layout:type:sqlcmd_operator]
# NOTE: This is the spacing between the operator and the colon
spacing_before = touch

[sqlfluff:layout:type:slash]
spacing_before = any
spacing_after = any

[sqlfluff:layout:type:comment]
spacing_before = any
spacing_after = any

[sqlfluff:layout:type:pattern_expression]
# Snowflake pattern expressions should not have their spacing changed.
spacing_within = any

[sqlfluff:layout:type:bind_variable]
spacing_within = touch

[sqlfluff:layout:type:placeholder]
# Placeholders exist "outside" the rendered SQL syntax
# so we should not enforce any particular spacing around
# them.
spacing_before = any
spacing_after = any

[sqlfluff:layout:type:common_table_expression]
# The definition part of a CTE should fit on one line where possible.
# For users which regularly define column names in their CTEs they
# may which to relax this config to just `single`.
spacing_within = single

# By setting a selection of clauses to "alone", we hint to the reflow
# algorithm that in the case of a long single line statement, the
# first place to add newlines would be around these clauses.
# Setting this to "alone:strict" would always _force_ line breaks
# around them even if the line is not too long.
[sqlfluff:layout:type:select_clause]
line_position = alone

[sqlfluff:layout:type:where_clause]
line_position = alone

[sqlfluff:layout:type:from_clause]
line_position = alone

[sqlfluff:layout:type:join_clause]
line_position = alone

[sqlfluff:layout:type:groupby_clause]
line_position = alone

[sqlfluff:layout:type:orderby_clause]
line_position = alone

[sqlfluff:layout:type:having_clause]
line_position = alone

[sqlfluff:layout:type:limit_clause]
line_position = alone

# Template loop tokens should not dictate spacing around them.
[sqlfluff:layout:type:template_loop]
spacing_before = any
spacing_after = any

[sqlfluff:templater]
unwrap_wrapped_queries = True

[sqlfluff:templater:jinja]
apply_dbt_builtins = True

# Some rules can be configured directly from the config common to other rules
[sqlfluff:rules]
allow_scalar = True
single_table_references = consistent
unquoted_identifiers_policy = all

[sqlfluff:rules:capitalisation.keywords]
# Keywords
capitalisation_policy = upper
# Comma separated list of words to ignore for this rule
ignore_words = None
ignore_words_regex = None

[sqlfluff:rules:capitalisation.identifiers]
# Unquoted identifiers
extended_capitalisation_policy = upper
# Comma separated list of words to ignore for this rule
ignore_words = None
ignore_words_regex = None

[sqlfluff:rules:capitalisation.functions]
# Function names
extended_capitalisation_policy = upper
# Comma separated list of words to ignore for this rule
ignore_words = None
ignore_words_regex = None

[sqlfluff:rules:capitalisation.literals]
# Null & Boolean Literals
capitalisation_policy = upper
# Comma separated list of words to ignore for this rule
ignore_words = None
ignore_words_regex = None

[sqlfluff:rules:capitalisation.types]
# Data Types
extended_capitalisation_policy = upper
# Comma separated list of words to ignore for this rule
ignore_words = None
ignore_words_regex = None

[sqlfluff:rules:ambiguous.join]
# Fully qualify JOIN clause
fully_qualify_join_types = both

[sqlfluff:rules:ambiguous.column_references]
# GROUP BY/ORDER BY column references
group_by_and_order_by_style = consistent

[sqlfluff:rules:aliasing.table]
# Aliasing preference for tables
aliasing = explicit

[sqlfluff:rules:aliasing.column]
# Aliasing preference for columns
aliasing = explicit

[sqlfluff:rules:aliasing.unused]
alias_case_check = dialect

[sqlfluff:rules:aliasing.length]
min_alias_length = None
max_alias_length = None

[sqlfluff:rules:aliasing.forbid]
# Avoid table aliases in from clauses and join conditions.
# Disabled by default for all dialects unless explicitly enabled.
# We suggest instead using aliasing.length (AL06) in most cases.
force_enable = False

[sqlfluff:rules:convention.not_equal]
# Consistent usage of preferred "not equal to" comparison
preferred_not_equal_style = ansi

[sqlfluff:rules:convention.select_trailing_comma]
# Trailing commas
select_clause_trailing_comma = forbid

[sqlfluff:rules:convention.count_rows]
# Consistent syntax to count all rows
prefer_count_1 = False
prefer_count_0 = False

[sqlfluff:rules:convention.blocked_words]
# Comma separated list of blocked words that should not be used
blocked_words = None
blocked_regex = None
match_source = False

[sqlfluff:rules:convention.casting_style]
# SQL type casting
preferred_type_casting_style = shorthand

[sqlfluff:rules:references.qualification]
subqueries_ignore_external_references = True

[sqlfluff:rules:references.consistent]
force_enable = False

[sqlfluff:rules:references.keywords]
# Keywords should not be used as identifiers.
unquoted_identifiers_policy = aliases
quoted_identifiers_policy = none

[sqlfluff:rules:references.special_chars]
# Special characters in identifiers
unquoted_identifiers_policy = all
quoted_identifiers_policy = all
allow_space_in_identifier = False
additional_allowed_characters = None

[sqlfluff:rules:references.quoting]
# Policy on quoted and unquoted identifiers
prefer_quoted_identifiers = False
prefer_quoted_keywords = False
case_sensitive = True

[sqlfluff:rules:layout.long_lines]
# Line length
ignore_comment_lines = False
ignore_comment_clauses = False

[sqlfluff:rules:layout.select_targets]
wildcard_policy = single

[sqlfluff:rules:structure.subquery]
# By default, allow subqueries in from clauses, but not join clauses
forbid_subquery_in = both

[sqlfluff:rules:structure.join_condition_order]
preferred_first_table_in_join_clause = earlier
```

### Are you willing to work on and submit a PR to address the issue?

- [x] Yes I am willing to submit a PR!

### Code of Conduct

- [x] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by peterbud** [comment]
Can you remove the 'AR.' from this line:
```sql
...
 DC.VALUE::VARCHAR AS AR.DIAG_CODE
...
```

and use this:
```sql
...
 DC.VALUE::VARCHAR AS DIAG_CODE
...
```

**Comment by ahwitz** [comment]
Same error, unfortunately, I just lazily pasted that in when trying to remove unrelated errors. The `AR.` prefix in the next line also was extraneous.

Updated it in the original comment (and triple-tick'd the config); updated output:

```
==== finding fixable violations ====
WARNING    Fixes for LT02 not applied, as it would result in an unparsable file. Please report this as a bug with a minimal query which demonstrates this warning.
WARNING    Fixes for LT09 not applied, as it would result in an unparsable file. Please report this as a bug with a minimal query which demonstrates this warning.
== [models/final/test.sql] FAIL
L:   2 | P:   5 | LT09 | Select targets should be on a new line unless there is
                       | only one select target. [layout.select_targets]
L:   2 | P:  11 | LT02 | Expected line break and indent of 8 spaces before 'ID'.
                       | [layout.indent]
L:  10 | P:   5 | LT09 | Select targets should be on a new line unless there is
                       | only one select target. [layout.select_targets]
L:  10 | P:  20 | LT02 | Expected line break and indent of 8 spaces before 'GI'.
                       | [layout.indent]
L:  16 | P:   1 | LT02 | Expected indent of 4 spaces. [layout.indent]
==== no fixable linting violations found ====
All Finished!
  [8 unfixable linting violations found]
All Finished
  [1 templating/parsing errors found]
```

**Comment by peterbud** [comment]
OK, first I wanted to make sure it parses at least - and it wasn't even parsing with the `AR.` prefix.

**Comment by peterbud** [comment]
Cannot be reproduced with the latest `sqlfluff`. It fixes the file. Also you should update the syntax of the `.sqlfluff` config file.

```bash
> sqlfluff fix test.sql 

WARNING: Config file /home/.../.sqlfluff set a deprecated config value `indentation:allow_implicit_indents`. This will be removed in a later release. This has been mapped to `indentation:implicit_indents` set to a value of 'allow' for this run. Please update your configuration to remove this warning. 

The allow_implicit_indents config has been replaced by implicit_indents. Use 'forbid' (was false), 'allow' (was true), or 'require' instead.

See https://docs.sqlfluff.com/en/stable/perma/configuration.html for more details.

==== finding fixable violations ====
WARNING    
WARNING: Config file /home/.../.sqlfluff set a deprecated config value `indentation:allow_implicit_indents`. This will be removed in a later release. This has been mapped to `indentation:implicit_indents` set to a value of 'allow' for this run. Please update your configuration to remove this warning. 

The allow_implicit_indents config has been replaced by implicit_indents. Use 'forbid' (was false), 'allow' (was true), or 'require' instead.

See https://docs.sqlfluff.com/en/stable/perma/configuration.html for more details.
 
== [test.sql] FAIL                                                                                                            
L:   2 | P:  11 | LT02 | Expected line break and indent of 8 spaces before 'ID'.
                       | [layout.indent]
L:  10 | P:  20 | LT02 | Expected line break and indent of 8 spaces before 'GI'.
                       | [layout.indent]
L:  16 | P:   1 | LT02 | Expected indent of 4 spaces. [layout.indent]
== [test.sql] FIXED
3 fixable linting violations found
  [3 unfixable linting violations found]
```

Result:
```sql
WITH GROUPED_IDS AS (
    SELECT
        ID
        , FIRST_VALUE(ID) OVER (
            PARTITION BY GROUP_KEY ORDER BY (DATA_SOURCE = '[removed for bug submission]', DATA_SOURCE) DESC
        ) AS GROUPED_ID
    FROM BASE_GROUPED
)

, DIAGNOSIS_CODES AS (
    SELECT DISTINCT
        GI.GROUPED_ID AS ID
        , DC.VALUE::VARCHAR AS DIAG_CODE
        , IFF(CONTAINS(DIAG_CODE, ','), NULL, DIAG_DESCRIPTION) AS DIAG_DESCRIPTION
    FROM GROUPED_IDS AS GI
    INNER JOIN ADT_RAW AS AR
        ON GI.ID = AR.ID
    , LATERAL FLATTEN(INPUT => SPLIT(AR.DIAG_CODE, ','), OUTER => TRUE) AS DC
)

SELECT *
FROM ADT_RAW

```

**Comment by ahwitz** [comment]
Great to hear - will forward this on to DBT support and ask when they're gonna update the bundled version. Thanks a ton, sorry for wasting your time!
